lib/srv/app/tcpserver.go

```diff
@@ -81,3 +96,36 @@ func (s *tcpServer) handleConnection(ctx context.Context, clientConn net.Conn, i
 	}
 	return nil
 }
+
+// dialMultiPortTCPApp assumes that app has TCP ports specifies and dials targetPort if it's found
+// in TCP ports.
+//
+// If the client did not include targetPort (it's equal to zero), it dials the first port found in
+// TCP ports.
+func dialMultiPortTCPApp(ctx context.Context, dialer net.Dialer, appAddr *utils.NetAddr, targetPort int, app apitypes.Application) (net.Conn, error) {
+	// If the client didn't supply a target port, use the first port found in TCP ports. This is to
+	// provide backwards compatibility.
+	//
+	// In theory, this behavior could be removed in the future if we guarantee that all clients always
+	// send a target port when connecting to multi-port apps, but no such effort was undertaken so far.
+	if targetPort == 0 {
+		firstPort := int(app.GetTCPPorts()[0].Port)
```

Comment:
This is sort of spooky, we're only checking for `app.GetTCPPorts()` being nonempty in the caller so if we change it in the wrong way we might end up with a panic here. The logic is also a bit too spread out, perhaps we can rework it as something like this instead?

```go
var dialTarget string
switch {
	case len(app.GetTCPPorts()) < 1:
		// dialTarget = addr.String
	case targetPort == 0:
		// dialTarget = joinhostport(host, firstport)
	default:
		// check for port in range then dialTarget = joinhostport(host, port)
}
dialer.Dial(dialTarget)
```

Reply:
Also, should we error out if there's a target port but the app is not multi-port?

Reply:
> Also, should we error out if there's a target port but the app is not multi-port?

I thought about it when the RFD was in review and this is what I said back then (https://github.com/gravitational/teleport/pull/46169#discussion_r1768436228):

> Getting a cert for a multi-port app which then becomes single-port means that the cert allows access to the port from `uri` in the app spec, no matter what `RouteToApp.TargetPort` in the cert says.
> 
> We could technically disallow access in that case too. But given that we don't plan to do RBAC on ports for now, I think this behavior is okay.

Nowadays I'd say that we should error in this scenario. Imagine you set up a local proxy for port 1337, only for the admin to change the app and now all your connections through this proxy go to port 4242 instead. That would be more confusing than the connection getting dropped and the admin being able to recover error message from cluster logs.

Buuut, implementing this means that there's also the possibility of `TargetPort` matching the port in the URI after an app is no longer multi-port. From the UX standpoint, I think we should allow such connection to go through as it's still routed to the correct port.

Reply:
This is fine IMO.